### PR TITLE
[SPARK-34674][CORE][K8S] Close SparkContext after the Main method has finished

### PR DESCRIPTION
**Pull request by kotlovs**
### What changes were proposed in this pull request?
Close SparkContext after the Main method has finished, to allow SparkApplication on K8S to complete.
This is fixed version of [merged and reverted PR](https://github.com/apache/spark/pull/32081).

### Why are the changes needed?
if I don't call the method sparkContext.stop() explicitly, then a Spark driver process doesn't terminate even after its Main method has been completed. This behaviour is different from spark on yarn, where the manual sparkContext stopping is not required. It looks like, the problem is in using non-daemon threads, which prevent the driver jvm process from terminating.
So I have inserted code that closes sparkContext automatically.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manually on the production AWS EKS environment in my company.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Thank you, @kotlovs .

cc @mridulm , too

**Review by dongjoon-hyun — APPROVED** [review]
The main body is the same with the once merged one and the only change is the added `if` statement.
+1, LGTM (if it passes all CIs).

**Comment by dongjoon-hyun** [comment]
Jenkins also passed Java/Scala/Python tests and currently running irrelevant R test.
Merged to master/3.1. Thank you for your patience and persistence, @kotlovs !

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:956]: Can we check if we're in K8s here ?

Comment (mridulm) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:956]: +1 to making it k8s specific ...
Ideally, we should be using daemon threads for most of our background work ...

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:956]: +1

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:956]: Since SPARK-34674 was released already via Apache Spark 3.1.2, I will make another JIRA for this, @HyukjinKwon , @mridulm , @Ngone51 .

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:956]: Here is the PR.
- https://github.com/apache/spark/pull/33403

**Comment by attilapiros** [comment]
I think we need a config to control this feature otherwise we run into problems on spark job servers running on K8S.

For details please check https://github.com/apache/spark/pull/39775

**Pull request closed without merging**
